research/delf/delf/python/training/global_features/train.py

```diff
@@ -0,0 +1,719 @@
+  # ground truth files are downloaded.
+  for dataset in FLAGS.test_datasets.split(','):
+    if dataset not in _TEST_DATASET_NAMES:
+      raise ValueError('Unsupported or unknown test dataset: {}.'.format(
+              dataset))
+
+    data_dir = os.path.join(FLAGS.data_root, 'gnd_{}.pkl'.format(dataset))
+    if not os.path.isfile(data_dir):
+      raise ValueError(
+              '{} ground truth file at {} not found. Please download it '
+              'according to '
+              'the DELG instructions.'.format(dataset, FLAGS.data_root))
+
+  # Check if train dataset is downloaded and dowload it if not found.
+  sfm120k.download_train(FLAGS.data_root)
+
+  # Creating model export directory if it does not exist.
+  model_directory = global_features_utils.create_model_directory(
+          FLAGS.training_dataset, FLAGS.arch, FLAGS.pool, FLAGS.whitening,
+          FLAGS.pretrained, FLAGS.loss, FLAGS.loss_margin, FLAGS.optimizer,
+          FLAGS.lr,
+          FLAGS.weight_decay, FLAGS.neg_num, FLAGS.query_size, FLAGS.pool_size,
+          FLAGS.batch_size, FLAGS.update_every, FLAGS.image_size,
+          FLAGS.directory)
+
+  # Setting up logging directory, same as where the model is stored.
+  logging.get_absl_handler().use_absl_log_file('absl_logging', model_directory)
+
+  # Set cuda visible device.
+  os.environ['CUDA_VISIBLE_DEVICES'] = FLAGS.gpu_ids
+  global_features_utils.debug_and_log('>> Num GPUs Available: {}'.format(
+          len(tf.config.experimental.list_physical_devices('GPU'))),
+          FLAGS.debug)
+
+  # Allow memory growth.
+  gpus = tf.config.experimental.list_physical_devices('GPU')
+  if gpus:
+    try:
+      for gpu in gpus:
+        tf.config.experimental.set_memory_growth(gpu, True)
+    except RuntimeError as e:
+      print(e)
+
+  # Set random seeds.
+  tf.random.set_seed(0)
+  np.random.seed(0)
+
+  # Initialize the model.
+  if FLAGS.pretrained:
+    global_features_utils.debug_and_log(
+            '>> Using pre-trained model \'{}\''.format(FLAGS.arch))
+  else:
+    global_features_utils.debug_and_log(
+            '>> Using model from scratch (random weights) \'{}\'.'.format(
+                    FLAGS.arch))
+
+  model_params = {'architecture': FLAGS.arch, 'pooling': FLAGS.pool,
+                  'whitening': FLAGS.whitening, 'pretrained': FLAGS.pretrained,
+                  'data_root': FLAGS.data_root}
+  model = global_model.GlobalFeatureNet(**model_params)
+  global_features_utils.debug_and_log('>> Network initialized.')
+
+  global_features_utils.debug_and_log('>> Loss: {}.'.format(FLAGS.loss))
+  # Define the loss function.
+  if FLAGS.loss == 'contrastive':
+    criterion = ranking_losses.ContrastiveLoss(margin=FLAGS.loss_margin)
+  elif FLAGS.loss == 'triplet':
+    criterion = ranking_losses.TripletLoss(margin=FLAGS.loss_margin)
+  else:
+    raise RuntimeError('Loss {} not available!'.format(FLAGS.loss))
+
+  # Defining parameters for the training.
+  start_epoch = 1
```

Comment:
nit: is this variable 1-indexed? Maybe it should start at 0?

Reply:
Sometimes (for example when pre-computing whitening) we run evaluation before the network training and then we set the epoch to 0. In other cases, we start from 1. Should I change that?

Reply:
No, I think it should be fine. Maybe you can explain this in the comments / descriptions (and later in the README for using this training script)